sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/kllAggregates.scala

```diff
@@ -0,0 +1,405 @@
+import org.apache.spark.sql.catalyst.trees.UnaryLike
+import org.apache.spark.sql.errors.QueryExecutionErrors
+import org.apache.spark.sql.types._
+
+/** This represents the aggregation buffer for the below aggregate functions. */
+case class KllBufferBigint(sketch: KllLongsSketch = KllLongsSketch.newHeapInstance()) {
+  def serialize(): Array[Byte] = sketch.toByteArray
+  def eval(): Array[Byte] = sketch.toByteArray
+}
+case class KllBufferFloat(sketch: KllFloatsSketch = KllFloatsSketch.newHeapInstance()) {
+  def serialize(): Array[Byte] = sketch.toByteArray
+  def eval(): Array[Byte] = sketch.toByteArray
+}
+case class KllBufferDouble(sketch: KllDoublesSketch = KllDoublesSketch.newHeapInstance()) {
+  def serialize(): Array[Byte] = sketch.toByteArray
+  def eval(): Array[Byte] = sketch.toByteArray
+}
+
+/**
+ * The KllSketchAggBigint function utilizes an Apache DataSketches KllLongsSketch instance to
+ * compute quantiles of the values of an input expression (such as an input column in a table).
+ * It outputs the binary representation of the KllLongsSketch.
+ *
+ * See [[https://datasketches.apache.org/docs/KLL/KLLSketch.html]] for more information.
+ *
+ * @param child
+ *   child expression against which quantile computation will occur
+ * @param mutableAggBufferOffset
+ *   offset for mutable aggregation buffer
+ * @param inputAggBufferOffset
+ *   offset for input aggregation buffer
+ */
+// scalastyle:off line.size.limit
+@ExpressionDescription(
+  usage = """
+    _FUNC_(expr) - Returns the KllLongsSketch compact binary representation.
+  """,
+  examples = """
+    Examples:
+      > SELECT LENGTH(kll_sketch_to_string_bigint(_FUNC_(col))) > 0 FROM VALUES (1), (2), (3), (4), (5) tab(col);
+       true
+  """,
+  group = "agg_funcs",
+  since = "4.1.0")
+// scalastyle:on line.size.limit
+case class KllSketchAggBigint(
+    override val child: Expression,
+    override val mutableAggBufferOffset: Int = 0,
+    override val inputAggBufferOffset: Int = 0)
+    extends TypedImperativeAggregate[KllBufferBigint]
+    with UnaryLike[Expression]
+    with ExpectsInputTypes {
+  def this(child: Expression) = this(child, 0, 0)
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): KllSketchAggBigint =
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): KllSketchAggBigint =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+  override protected def withNewChildInternal(newInput: Expression): KllSketchAggBigint =
+    copy(child = newInput)
+
+  override def dataType: DataType = BinaryType
+  override def inputTypes: Seq[AbstractDataType] =
+    Seq(
+      TypeCollection(
+        ByteType,
+        IntegerType,
+        LongType,
+        ShortType))
+  override def nullable: Boolean = false
+  override def prettyName: String = "kll_sketch_agg_bigint"
+  override def createAggregationBuffer(): KllBufferBigint = KllBufferBigint()
```

Comment:
I'd recommend we add support here for different size heaps for each of the aggregation buffers. The given code will only support the default size of the sketch. Similar to what was done with Theta sketches and HLL. Users would benefit from this being configurable.

The constructor: https://github.com/apache/datasketches-java/blob/main/src/main/java/org/apache/datasketches/kll/KllFloatsSketch.java#L85

Reply:
Sure, this is done.

@cboumalh Looking at the Theta sketches implementation, are we actually checking that the corresponding argument is constant-foldable before we `.eval` it? If not, we could get a weird error if it's not constant. 

[1] https://github.com/apache/spark/blob/6d5f111cc86a0194130d77510a4d739e32c875df/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/thetasketchesAggregates.scala#L97